### ISSUE TITLE
mariadb ubuntu16.04 g++

### ISSUE THREAD

**Issue by fyt000** [liancheng/amy#20]
`g++ -DUSE_BOOST_ASIO=1 --std=c++17 -lmariadbclient -lpthread -lboost_system -I$AMY_HOME/include mariadb_async_single_query.cpp utils.cpp -Wl,--verbose`
Couldn't get it to work on g++, getting spews of linking errors. This is on ubuntu16.04, `g++ (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0`, self compiled boost 1.68 and `apt-get install libmariadbclient-dev`. Note that it works fine if clang++ is used instead.

some of the errors I got
``ariadb_async_single_query.cpp:(.text+0x88a): undefined reference to `mysql_get_socket'
/tmp/ccrbcwld.o: In function `void amy::(anonymous namespace)::async_wait_mysql<amy::mariadb_service::query_handler<std::_Bind<void (*(amy::placeholders::_ph_error, std::reference_wrapper<amy::basic_connector<amy::mariadb_service> >))(boost::system::error_code const&, amy::basic_connector<amy::mariadb_service>&)> >::state, amy::mariadb_service::query_handler<std::_Bind<void (*(amy::placeholders::_ph_error, std::reference_wrapper<amy::basic_connector<amy::mariadb_service> >))(boost::system::error_code const&, amy::basic_connector<amy::mariadb_service>&)> > >(int&, amy::mariadb_service::query_handler<std::_Bind<void (*(amy::placeholders::_ph_error, std::reference_wrapper<amy::basic_connector<amy::mariadb_service> >))(boost::system::error_code const&, amy::basic_connector<amy::mariadb_service>&)> >::state&, amy::mariadb_service::query_handler<std::_Bind<void (*(amy::placeholders::_ph_error, std::reference_wrapper<amy::basic_connector<amy::mariadb_service> >))(boost::system::error_code const&, amy::basic_connector<amy::mariadb_service>&)> >&)':``

``mariadb_async_single_query.cpp:(.text._ZN5boost4asio6detail22deadline_timer_serviceINS1_18chrono_time_traitsINSt6chrono3_V212steady_clockENS0_11wait_traitsIS6_EEEEE7destroyERNSA_19implementation_typeE[_ZN5boost4asio6detail22deadline_timer_serviceINS1_18chrono_time_traitsINSt6chrono3_V212steady_clockENS0_11wait_traitsIS6_EEEEE7destroyERNSA_19implementation_typeE]+0x29): undefined reference to `boost::system::detail::system_category_instance'``

Also played around with link ordering, didn't help.

Few other notes, mariadb stuff needs c++14 due to the use of `make_unique`. On arch/manjaro where mysql is basically mariadb linking should be -lmysqlclient. And g++ works fine on arch/manjaro.

**Comment by ghost** [comment]
* `undefined reference to mysql_get_socket'` :   you didn't link the the mariadbclient lib. 
 Solution :  reinstall the libmariadbclient-dev by https://downloads.mariadb.org/mariadb/repositories/#mirror=tuna&distro=Ubuntu&distro_release=xenial--ubuntu_xenial&version=10.0
```
sudo apt-get install software-properties-common
sudo apt-key adv --recv-keys --keyserver hkp://keyserver.ubuntu.com:80 0xF1656F24C74CD1D8
sudo add-apt-repository 'deb [arch=amd64,arm64,i386,ppc64el] http://mirrors.tuna.tsinghua.edu.cn/mariadb/repo/10.0/ubuntu xenial main'
sudo apt update
sudo apt install libmariadbclient-dev
```
* `undefined reference to boost::system::detail::system_category_instance'`: This is a common error of using different c++ standard or other flags in a project. 
 Solution 1. use boost_system header-only by define `BOOST_ERROR_CODE_HEADER_ONLY`
`g++-7 mariadb_async_single_query.cpp utils.cpp -I$BOOST_HOME -I$AMY_HOME/include -lmysqlclient -lpthread -DUSE_BOOST_ASIO=1 -DBOOST_ERROR_CODE_HEADER_ONLY`
 Solution 2. recompile boost_system with the same C++ standard or other flags.

**Comment by fyt000** [comment]
Thanks for replying.

`-DBOOST_ERROR_CODE_HEADER_ONLY` this works great.

Tried installing lastest mariadbclient
`The following packages have unmet dependencies:
 libmariadbclient-dev : Depends: libmariadbclient18 (= 1:10.1.34-0ubuntu0.18.04.1) but 1:10.3.10+maria~bionic is to be installed`
So it seems there is no point for me to use the added apt-repository.... the one bundled in ubuntu is already 10.1.34(> than the 10.0 you suggested). Also as noted, clang++ seems to be fine with the default ubuntu repo's libmariadbclient-dev. I tried building the c-connector myself and got a whole lot more linking problems. My friend on mac is having issues with missing mysql error code defines. He installed `mariadb-connector-c` via brew. It seems such a hassle to get the the connector library to link. 

Thanks again for helping but we are probably gonna try another postgres&asio library to see if its easier to build instead.
